doc/gnarwl.man
.TH GNARWL _MANSEC_
.SH NAME 
gnarwl \- GNU Neat Autoreply With LDAP

.SH SYNOPSIS
.B gnarwl 
[\-h]
[\-c\ <cfgfile>]
[\-a\ <address>]
[\-s\ <address>]

.SH DESCRIPTION
.B gnarwl 
is an email autoresponder, intended to be a successor to the old
vaction(1) program. Since a modern mailserver, usually serves hundreds 
(or even thousands) of mailaccounts, it is not sensible to give
(untrusted) users shell access so they may create/maintain the .forward file,
required by 
.B vacation(1).
.P
With 
.B gnarwl
, all user-suplied data is stored within an LDAP database, so 
there are no per user ".forward" files (or even homedirs) needed.
Configuration is conveniently done via one systemwide configfile. 
.P
Like the old 
.B vacation(1)
program, gnarwl accepts incoming mail through
stdin, and will send outgoing mail via an external MTA (it even maintains
basic commandline compatibility, so it may be used as a drop in
replacement).
.P
Several gdbm databases are maintained, in order to make sure, a) mail does not
bounce back and force between gnarwl and another automated MUA, b) mailing
lists will not be bothered and c) specific local addresses may never produce 
automatic replies.
All these database files may be managed using the 
.B damnit(_MANSEC_)
program.

.P
.SH OPTIONS
.IP "-c <cfgfile>"
Use a different configfile than the one, compiled in.

.IP "-a <alias>"
Force <address> as receiving address.

.IP "-s <address>"
Force <address> as sending address.

.IP -h 
Print usage information.

.SH CONFIGURATION
.B gnarwl
typically uses one global configurationfile, but a per user setup is also
possible using the -c commandline switch. The following keywords are
recognized in the configfile:

.IP "map_sender <macroname>"
Binds a macroname (case insensitive), referring to the sender of an incoming 
email. Defaults to "$sender".

.IP "map_receiver <macroname>"
Binds a macroname (case insensitive), referring to the receiver(s) of an 
incoming email. Defaults to "$receiver".

.IP "map_subject <macroname>"
Binds a macroname (case insensitive), referring to the subject of an incoming 
email. Defaults to "$subject".

.IP "map_field <ldapattribute> <macroname>"
Binds a macroname (case insensitive), referring to a field in the resultset, 
returned by the database. There are no defaults for this directive.

.IP "server_uri <ldap_uri [ldap_uri ...]>"
URI of the databaseserver to query. Multiple URIs can be specified. No default value.

.IP "starttls <0|1>"
Enforce StartTLS on connect. Defaults to 0.

.IP "ca_cert </path/to/certificate>"
Use specified CA cert to validate ldap server certificate. No default value.

.IP "server <address>"
Address of the databaseserver to query. Defaults to localhost.

.IP "port <portnumber>"
Port, the LDAP server listens on. Defaults to 389.

.IP "scope <base|one|sub>"
The scope used for searching the database. Default is "sub".

.IP "login <string>"
Destinguished name to bind with to the LDAP database. Default is to bind 
anonymously.

.IP "password <string>"
Password to use for binding to the LDAP database. If a password is required
to access the server, then the configfile should belong to the gnarwl user 
and have file permission 0400.

.IP "base <destinguished name>"
Entrypoint of the search. There is no default for this directive, it must
be supplied by the administrator.

.IP "protocol <2|3>
Select protocol to bind to the ldapserver. The default is 3.

.IP "queryfilter <ldap queryfilter>"
Search pattern to match against the database. Defaults to:
"(&(mail=$recepient)(vacationActive=TRUE)".

.IP "result <ldap attribute>"
The name of the attribute, that is to be taken as the emailbody. The content
of this field will be pasted in between the data found via forceheader and 
forcefooter directives. Afterwards all remaining macros are expanded in the
order of declaration, and the result will be piped through to the MTA.

.IP "blockfiles <directory>"
The directory, where gnarwl stores it's blockfiles. These files are
required to keep track on who was sent an automatic reply. Default is:
"_HOMEDIR_/block/".

.IP "umask <mode>
What permission to give newly generated database files. The default is
0600.

.IP "blockexpire <number>"
How long (in hours) to block a certain sender/recepient combo. Default is
48 hours. Setting <number> to 0 disables the feature (not recommended). No
blockfiles are read/written in this case.

.IP "maxreceivers <number>"
Ignore incoming email, specifying too many receiving addresses. It does
not matter, whether these are local or not, as 
.B gnarwl 
doesn't know domains. Default is 256.

.IP "maxheader <number>"
Ignore incoming email with more than this number of header lines. Lines are
counted before unfolding them, so a folded line really counts as at least
two lines. Default is 256.

.IP "badheaders <filename>"
Path to a database file, containing matching patterns for the mailheader.
If an entry stored in this file matches a line in the header exactly, then
this mail will be ignored by
.B gnarwl
\&. (useful to avoid sending automatic replies to mailing lists). This feature
is deactivated by default.

.IP "blacklist <filename>"
Pointer to a database file, containing emailaddresses, 
.B gnarwl
is not allowed
to generate automatic replies for (useful to prevent automatic replies from 
addresses, which are shared by several people). This feature is deactivated 
by default.

.IP "forceheader <filename>"
Path to a text file, containing a standardized header, that is to be 
pasted in front of every outgoing mail. This file should end with a single
empty line. Otherwise it is assumed, that the users are allowed to
continue the header and will provide the separating empty line themselves.
Default is not to force anything (that is: The user has to supply the header
in the "result" attribute).

.IP "forcefooter <filename>"
Path to a text file, containing a standardized footer, that is to be appended
at the end of every generated mail. Default is to not to force anything.

.IP "mta <prog> [<args>]"
Specify MTA for sending mail. It must be able to accept mail on STDIN.
Default is "/usr/sbin/sendmail".

.IP "charset <encoding>"
LDAP stores text in unicode, which is ok, as long as outgoing mail doesn't
contain any non ASCII characters. However, locale specific characters (like
german umlaute) end up as strange glyphs. With the "charset" directive,
gnarwl tries to convert these to the correct symbols. The <encoding> argument
must contain a string recognized by iconv(3).
Default is not to try to convert anything (assume US-ASCII charset / MIME
encoded mail).

.IP "recvheader <string>"
A whitespace separated list of headers (case does not matter), which may
contain receiving addresses. Defaults to: "To Cc".

.IP "loglevel <0|1|2|3>"
Specifies what to send to the syslog. A higher loglevel
automatically includes all lower loglevels (see section syslog for more 
information).

.IP "deref <never|search|find|always>"
Controls what LDAP alias dereferencing will be performed. Default is "find".

.SH SYSLOG
Since 
.B gnarwl
is not meant to be invoked by anything but the mailsystem,
it'll never print out messages to the systemconsole, but logs them via 
syslog(3), using the facility "mail". A log line is always of the following
format:
.P
<level>/<origin> <message>
.P
The <level> field indicates the severity of the message, it corresponds to the
"loglevel" config directive. Possible values are:
.P
.IP "CRIT (loglevel 0)"
Critical messages. 
.B gnarwl cannot continue and will die with a non-zero exit 
code. This usually causes the mailsystem to bounce mail.

.IP "WARN (loglevel 1)"
A warning. 
.B gnarwl 
can will continue, but not with the full/intended functionality.

.IP "INFO (loglevel 2)"
Status information. A message in the INFO loglevel indicates normal
behaviour.

.IP "DEBUG (loglevel 3)" 
Debugging information. 
.B gnarwl will log a lot of information on how mail is processed. 

.P
The <origin> field gives a short hint about what caused the log entry in 
question, while <message> contains a short description of what actually
happened.


.SH AUTHOR
Patrick Ahlbrecht <p.ahlbrecht@billiton.de>

.SH SEE ALSO
.BR vacation (1),
.BR postfix (1),
.BR iconv (1),
.BR damnit (_MANSEC_),
.BR rfc822

.SH FILES
.I _CONFDIR_/gnarwl.cfg 
.RS 
main configuration file.
.RE
.P
.I _HOMEDIR_/.forward
.RS 
forward file for the mailsystem.
.RE
.P
.I _HOMEDIR_/blacklist.db
.RS
.B gnarwl 
won't send an autoreply for anyone whose emailaddress is listed
herin.
.RE
.P
.I _HOMEDIR_/badheaders.db
.RS
.B gnarwl 
will ignore mail, it is able to match a headerline with an entry in
this file. Case is significant, no wildcards are expanded.
.RE
.P
.I _HOMEDIR_/header.txt
.RS
Standard header to paste in front of every outgoing mail.
.RE
.P
.I _HOMEDIR_/footer.txt
.RS
Standard footer to append to every outgoing mail.
.RE
.P
.I _HOMEDIR_/block/*
.RS 
block files.
.RE
.P


doc/HISTORY
The versioning scheme of gnarwl consists of a major and a minor version
number. An increase in the former means a new feature was added, while an
increase in the later is to document a maintainance fix.

1.0:
  First public release.
  
1.1:
- Added a retrycounter and random sleep call to dbstuff.dbLock(). This should 
  fix an infinite loop problem of 1.0 where to instances of gnarwl would 
  try to access the same .db file over and over again, blocking each other.

- Updated documentation

- Added this file

- added README.pitfalls

- Renamed SECURITY README.security

- Updated Makefile (simplified it a bit)

- Removed hardcoded "sendmail" from  execl call in gnarwl.doReply(),
  replaced it with the result of basename(cfg.mta).

- openlog is now called once in main(), instead of immediatly before
  every syslog() call (what made me waste resources like this?).

- Removed exit(0) call in main(), return 0; schould do as well.  

- Updated gnarwl.printUsage(): Added version number, uses a single printf
  now.

- replaced exit(1) and exit(0) with exit(EXIT_FAILURE) and exit(EXIT_SUCCESS).
  Code should be more portable this way.

- gnarwl.addAddr() now checks first against cfg.maxmail before doing
  anything else.

- gnarwl.addAddr() now strips emailaddresses with angle brackets, e.g.:
  John Doe <john@doe.org> becomes john@doe.org
  
- Added new config switch "log_verbose", since -v cannot be specified in
  .forward files.

- closed some memory leaks


1.2
- Added a newline character to gnarwl.c (gcc 2.96 complained about it)
- Updates docs
  * fixed some typos
  * added a warning about openldap2.0.23 to INSTALL
  * added a manpage (bouyah!).
- Updated Makefile:
  * $(VER) is now computed from the directory basename, instead of being 
    hardcoded
  * added new variable $(CONFDIR)
  * The macros CFGFILE and VERSION are now defined via compiler switch
- Changed gnarwl.printUsage():
  * now also prints out default configfilename
  * versioninformation is taken from VERSION macro (instead of being
    hardcoded in the source).

2.0
- Renamed this file from CHANGES to HISTORY
- moved src/dbstuff.c to dbaccess.c
- killed dbaccess.dbExpired() ; integrated it's logic directly into dbCheck()
- added new management utility, damnit (yes, that's it's name).
- Splitted gnarwl.h -> every module has now it's own header
- Updates Makefile: Everything is now installed under /usr/local per default
- Changed gnarw.doReply() return type from int to void (since doReply always
  returned TRUE are bailed out itself if something went wrong).
- Reformulated most program messages
- More logging
- Base64 support. Actually it seems as if it was supported all the time, 
  but I didn't know of it #-) (So much for the documentation of openldap
  ;-). Why does ldapsearch return base64 encoded data, but 
  ldap_get_first_entry() not ?!).
- Changed return type of dbaccess.dbLock() from int to void. This method 
  will now exit(EXIT_FAILURE) directly if locking was unsucesfull.
- Fixed a bug in dbaccess.dbLock(): The do-while loop lacked the correct
  break condition, so it always looped 10 times.
- Replaced util.ewe(char*) with util.oom(). 
- Removed default away message (waste of space and didn't look pretty
  anyway).
- Added more filechecking to dbaccess.dbCheck(). It was rather weak
  before (would fail, if another process had the db file open for writing).
- Added flexible deny rules for mail.
- The blacklist is now stored in a gdbm file.
- Renamed bl_file to mail_blacklist
- Updated installation process
- Updated docs
- gnarwl.readLDAP() now uses ldap_init() instead of ldap_open()
- Added new version of billiton.schema
- Found a "nice feature": you may not free(2) the argument of basename(3).
  Fixed gnarwl.doReply accordingly. On the second thought: using basename
  was a stupid idea in the first place -> removed it completely.
- gnarwl is no longer bound to sendmail as MTA; MTA is now completly
  configureable (with parameters).
- renamed cfg.v_message to cfg.result
- added README.upgrade
- Updated manpage
- Disallowed autoreply for shared addresses.
- Renamed mail_limit to mail_limit_receivers, added mail_limit_header
- Merged README.pitfalls and README.security into FAQ
- Only the attributes specified in the config file will be queried
- Reworked gnarwl.readLDAP()
- added free(rep) after a call to gnarwl.doReply
- Removed v_begin and v_end, gnarwl now uses flexible macros
- Removed docs/USING
- gnarwl.readLDAP() now uses ldap_search_st instead of ldap_search_s
- Renamed config directive log_verbose to log_level.
- Eliminated "-v" commandlineswitch
- Added additional loglevels.
- Found a hefty bug in gnarwl.readHeader -> reimplemented it along with
  gnarwl.addAddr and ganrwl.main (they looked ugly anyway).
- added new function: dbaccess.dbOpen
- Moved declaration of "verbose" and "cfg" from gnarwl.c to conf.c, so
  damnit may link against the other objectfiles.
- Cleaned up the mess in the headerfiles:
  * Moved config structure from gnarwl.h to conf.h
  * All functions are now documentated in the corresponding headerfiles
  * gnarwl.h no longer includes all the other headers
- Added "contrib" branch for add-on software.

2.1
- Fixed a minor bug in dbaccess.dbCheck() and dbaccess.dbLock(): Both
  routines would ommit the trailing \0 of the key value. This was not
  critical, as both did so, but it was incompatible with damnit.
- Fixed a memory leak in damnit.listFile() (key.dptr and val.dptr are
  now freed).
- "make uninstall" will now also remove the system account of gnarwl (if
  present).
- Added a new macro "USERADD_ARG" to the Makefile.
- Updated docs
- Gnarwl.readHeader() would call util.cleanAddress without testing whether 
  the result was NULL or not (and therefore could segfault if one 
  supplied illegal from/to/cc or reply-to addresses) -> Added a test for
  NULL after every cleanAddress call.
- Changed places of main.doReply() and dbaccess.dbLock in gnarwl.main.

contrib/horde/gnarwl.php
<?php
/*
This is a little hack to add support for ldap/gnarwl based vacation in horde.
You have to adjust your vacation/config/conf.php and add these values:

$conf['server']['driver'] = 'gnarwl';
$conf['server']['params'] = array();
$conf['server']['params']['ldap_server'] = 'localhost';
$conf['server']['params']['ldap_base'] = 'ou=Users,dc=example,dc=de';
 
Of ocurse you have to adjust your LDAP server and base settings.
Copy this file in vacation/lib/Driver/gnarwl.php.


Horde can be found under http://www.horde.org.
Gnarwl can be found under http://www.oss.billiton.de/software.shtml

This script by C. Schwede <alt-f4@cschwede.de>, 2003.
*/


class Vacation_Driver_gnarwl extends Vacation_Driver {
    var $params;


function Vacation_Driver_gnarwl($params = array())
{
    $this->params = $params;
}


   

function check_config() {
    if (empty($this->params['ldap_server']) ||
        empty($this->params['ldap_base']) ) {
            $this->err_str = _("The module is not properly configured!");
            return false;
    }
    return true;
}


/*
 * @param string    $user       The username to enable vacation for.
 * @param string    $realm      The realm of the user - isn't used with gnarwl.
 * @param string    $pass       The password for the user.
 * @param string    $message    The message to install.
  
*/

function set_vacation($user, $realm, $pass, $message) 
        { if (!$this->check_config()) {return false;}
        $connect=ldap_connect($this->params['ldap_server']) or die($this->err_str = ("Can't connect to LDAP Server") );
        $bind=ldap_bind($connect, "uid=$user,".$this->params['ldap_base'], $pass);
                if ($bind) 
                        {
                                $info["vacationInfo"]=utf8_encode($message);
                                $info["vacationActive"]="TRUE";
                                $r=ldap_modify($connect,"uid=$user," . $this->params['ldap_base'], $info);
                                if (!$r) {$this->err_str = _(ldap_error($connect));return false;}
                        }
                                else {$this->err_str = _("Couldn't bind to LDAP server");return false;} 
                return true;
                
                ldap_close($connect);
        }


function unset_vacation($user, $realm, $pass, $message) 
        { if (!$this->check_config()) {return false;}
        $connect=ldap_connect($this->params['ldap_server']) or die($this->err_str = ("Can't connect to LDAP Server") );
        $bind=ldap_bind($connect, "uid=$user,".$this->params['ldap_base'], $pass);
                if ($bind) 
                        {
                                $info["vacationActive"]="FALSE";
                                $r=ldap_modify($connect,"uid=$user," . $this->params['ldap_base'], $info);
                                if (!$r) {$this->err_str = _(ldap_error($connect));return false;}
                        }
                                else {$this->err_str = _("Couldn't bind to LDAP server");return false;} 
                return true;
                
                ldap_close($connect);
        }



}
?>
